packages/@aws-cdk/cdk/lib/core/tokens.ts

```diff
@@ -518,3 +516,83 @@ function resolveFragment(fragment: Fragment): any {
 function regexQuote(s: string) {
   return s.replace(/[.?*+^$[\]\\(){}|-]/g, "\\$&");
 }
+
```

Comment:
I am wondering: if we stopped trying to make "resolve" document-agnostic and basically tie in "resolve" to CloudFormation template format, how much would it simplify this? Seems that we might be able to simplify dramatically and get rid of this options stack thing by baking in the knowledge of both CFN concatenation and cross stack reference resolution into `resolve` it self.